### PR TITLE
gh-110481: Fix Py_SET_REFCNT() integer overflow

### PR DESCRIPTION
**Pull request by vstinner**
If Py_NOGIL is defined and Py_SET_REFCNT() is called with a reference count larger than UINT32_MAX, make the object immortal.

Set _Py_IMMORTAL_REFCNT constant type to Py_ssize_t to fix the following compiler warning:

Include/internal/pycore_global_objects_fini_generated.h:14:24: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and 'unsigned int' [-Wsign-compare]

    if (Py_REFCNT(obj) < _Py_IMMORTAL_REFCNT) {
        ~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-110481 -->
* Issue: gh-110481
<!-- /gh-issue-number -->

### CODE REVIEW

**Comment by vstinner** [comment]
@colesbury @eduardo-elizondo @corona10: Would you mind to review these changes?

**Comment by eduardo-elizondo** [comment]
@vstinner I can do a thorough review this Saturday if no one gets to it before then!

**Review by colesbury — COMMENTED** [review]
The biased reference counting changes lgtm. 

This adds some C-style casts. What's the "rule" for using C-style casts vs. `_Py_STATIC_CAST`/`_Py_CAST` in this file?

**Comment by vstinner** [comment]
Oh, I forgot about _Py_CAST(). i introduced it to avoid C++ compiler warnings with the strict "no old C cast" warning. Sadly, my complicated C++ _Py_CAST() implementation had to be reverted. There are now some basic C++ tests in test_cppext, but it doesn't catch forgotten _Py_CAST().

**Comment by vstinner** [comment]
> This adds some C-style casts. What's the "rule" for using C-style casts vs.  _Py_STATIC_CAST/_Py_CAST in this file?

I added _Py_CAST() calls.

**Comment by vstinner** [comment]
I modified the PR that Py_SET_REFCNT() makes an object immortal if refcnt is larger than UINT32_MAX. PR based on my previous PR which adds "immortal" to the documentation glossary: https://github.com/python/cpython/pull/112180

I also added comments in Py_INCREF() to describe the behavior when the reference count in UINT32_MAX in Py_INCREF() (not in the 3rd code path which checks explicitly for `_Py_IsImmortal()`).

**Comment by vstinner** [comment]
@eduardo-elizondo:
> @vstinner I can do a thorough review this Saturday if no one gets to it before then!

Sure, I will wait for your review :-)

File: Doc/using/configure.rst [hunk 1]
```diff
@@ -287,6 +287,8 @@ General Options
 
    .. versionadded:: 3.11
 
+.. _free-threading-buid:
```
Comment (Eclips4) [inline Doc/using/configure.rst]: I don't sure, but it's looks like unrelated to this PR. However, it's missing letter `l` in word `build` :)

Comment (vstinner) [reply Doc/using/configure.rst]: Oops, fixed.

**Comment by vstinner** [comment]
I prepared the documentation change of this PR by adding the "immortal" term to the documentation glossary: https://github.com/python/cpython/pull/112180

**Comment by vstinner** [comment]
Oh, I missed an obvious typo :-( Sadly, the NoGIL jobs didn't run in the pre-commit CI. Maybe because I created the PR 2 weeks ago?

Anyway, PR #112595 should fix the buildbot.
